Auto test + deploy using Github Action

yne opened the pull request:
## Description
This PR add the following:
- git tag creation will publish on NPM
  - Automatically change NPM package version + build + test + publish (need to setup secret)
  - Github Release page can automatically generate a release note
- branch push or PullRequest will trigg a test + status report
  - easier/more secure merge review process
- Github Page deployment is done by github Action
  - Action are needed to generate the assets served from the web /dist/ url
  - Move web page related assets to `/docs/` folder
- Use ESLint + ESBuild
  - More popular/maintained solution than jshint + rollup
  - NPM now report 0 vulnerabilities :tada: (edit: NVM I had to roll back mocha, so we are back to 2 vuln)

## Motivation and Context

I've made this PR because it's a burden to
- Accept PR that possibly contain hidden malware in they minified dist files => /dist/ files are built on CI + published.
- Manually do the versioning in the package.json (only @Jolg42 can publish on NPM anyway) => any maintainer can now do it from the gituhub release page.
- Update the RELEASE.md => redirect to the Github Release page that contain auto-generated summaries
- Update the CONTRIBUTOR.md => redirect to the Github Contributor page that contain all contributor
- See a vulnerabilities warning each time I install opentype.js

## How Has This Been Tested?

I've forked the opentype.js repos as opentype.yne.js and deployed it on [NPM](https://www.npmjs.com/package/opentype.yne.js?activeTab=explore) and on [github page](https://yne.fr/opentype.yne.js/)

I've also tested the following developer use cases (after `npm i opentype.yne.js`) :

index.mjs
```js
import opentype from 'opentype.yne.js';
import { parse } from 'opentype.yne.js';

// the folowing import don't work:
// import mod from 'opentype.yne.js/dist/opentype.module/js';
// Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.

console.log(opentype, parse)
```

index.js
```js
const ot = require('opentype.yne.js');
console.log(ot)
```

```html
<!-- same as https://unpkg.com/opentype.yne.js -->
<script src="https://unpkg.com/opentype.yne.js/dist/opentype.js"></script>
<script>console.log(opentype.parse)</script>

<script type=module>
import { parse } from "https://unpkg.com/opentype.yne.js/dist/opentype.module.js"
console.log(parse);
</script>
```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] I did `npm run test` and all tests passed green (including code styling checks).
- [ ] I have added tests to cover my changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the **README** accordingly.

yne commented on docs/font-inspector.html:
```diff
@@ -5,7 +5,7 @@
     <meta name="description" content="A JavaScript library to manipulate the letterforms of text from the browser or node.js.">
     <meta charset="utf-8">
     <link rel="stylesheet" href="site.css">
-    <script type="text/javascript" src="dist/opentype.js"></script>
```
revert

yne commented on src/util.js:
```diff
@@ -6,30 +6,10 @@ function isNode() {
     return typeof window === 'undefined';
 }
 
-function nodeBufferToArrayBuffer(buffer) {
```
split commi

yne commented on test/opentypeSpec.js:
```diff
@@ -66,23 +56,6 @@ describe('opentype.js', function() {
         assert.equal(aGlyph.path.commands.length, 14);
     });
 
-    it('handles a parseBuffer error', function(done) {
```
re-add or explain

yne commented:
All API and naming are keept for backward compatibility.
If not, warn me.

ILOVEPIE commented on src/tables/sfnt.js:
```diff
@@ -4,23 +4,23 @@
+import gsub from './gsub.js';
+import meta from './meta.js';
+import colr from './colr.js';
+import cpal from './cpal.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on src/tiny-inflate@1.0.3.esm.js:
```diff
@@ -0,0 +1,397 @@
+/*
```
Why are we adding a static dependency on someone else's code? If we're adding WOFF2 decompression (which has already been decided should be handled by using an external library in concert with Opentype.js to reduce bloat for people who don't need the decompression) we should use NPM, not statically including a library into our code.

ILOVEPIE replied:
Need some clarification on this.

yne replied:
Sorry I missed that one.
The reason is that [tiny-inflate](https://github.com/foliojs/tiny-inflate/) API does not provide an ESModule (`import`/`export`) version.

Lucky for us there are ESM-compatible versions (see: [tiny-inflate-es](https://github.com/code4fukui/tiny-inflate-es)) but sadly it is not published on NPM

ILOVEPIE replied:
What is it used for though?

yne replied:
It is only used for the WOFF (1) table decompression
See this paragraph : https://www.w3.org/TR/WOFF/#DataTables

WOFF2 use a much better and much complex algorythm.

Connum replied:
I don't really like the static dependency either, but I see why you needed to do it.
We could take a look at https://github.com/101arrowz/fflate - it claims to be *way* faster than tiny-inflate, while being about the same size when only using inflation via tree shaking.

Connum replied:
Just tried it out, it's as easy as
`import { decompressSync as inflate } from 'fflate';`

Connum replied:
Let's handle that separately and go with the static dependency for now.

ILOVEPIE commented on src/tokenizer.js:
```diff
@@ -153,11 +153,11 @@ Tokenizer.prototype.composeRUD = function (RUDs) {
     const hasFAILObject = obj => (
         typeof obj === 'object' &&
-        obj.hasOwnProperty('FAIL')
+        Object.prototype.hasOwnProperty.call(obj, 'FAIL')
```
Why this change?

ILOVEPIE commented on src/tokenizer.js:
```diff
@@ -439,7 +439,7 @@ Tokenizer.prototype.getContextRanges = function(contextName) {
     const registeredContexts = this.registeredContexts;
     for (const contextName in registeredContexts) {
-        if (registeredContexts.hasOwnProperty(contextName)) {
+        if (Object.prototype.hasOwnProperty.call(registeredContexts, contextName)) {
```
Why these changes?

ILOVEPIE commented on src/types.js:
```diff
@@ -1,7 +1,7 @@
 // All OpenType fonts use Motorola-style byte ordering (Big Endian)
 
-import check from './check';
+import check from './check.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on src/types.js:
```diff
@@ -207,9 +207,9 @@ sizeOf.LONGDATETIME = constant(8);
-            v.charCodeAt(3)];
+        v.charCodeAt(1),
+        v.charCodeAt(2),
+        v.charCodeAt(3)];
```
Are we changing code style?

ILOVEPIE commented on src/types.js:
```diff
@@ -733,9 +733,9 @@ encode.INDEX = function(l) {
-                           data);
+        encode.OffSize(offSize),
+        encodedOffsets,
+        data);
```
See above.

ILOVEPIE commented on test/bidi.js:
```diff
@@ -1,6 +1,8 @@
-import { loadSync } from '../src/opentype';
+import Bidi from '../src/bidi.js';
+import { parse } from '../src/opentype.js';
+import { readFileSync } from 'fs';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/featureQuery.js:
```diff
@@ -1,7 +1,9 @@
-import { ContextParams } from '../src/tokenizer';
+import { parse } from '../src/opentype.js';
+import FeatureQuery from '../src/features/featureQuery.js';
+import { ContextParams } from '../src/tokenizer.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/font.js:
```diff
@@ -1,5 +1,7 @@
 import assert from 'assert';
-import { Font, Glyph, Path, loadSync } from '../src/opentype';
+import { Font, Glyph, Path, parse } from '../src/opentype.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/glyph.js:
```diff
@@ -1,13 +1,15 @@
 import assert  from 'assert';
-import { loadSync, Glyph, Path } from '../src/opentype';
+import { parse, Glyph, Path } from '../src/opentype.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/layout.js:
```diff
@@ -1,6 +1,6 @@
-import { Font, Path, Glyph } from '../src/opentype';
-import Layout  from '../src/layout';
+import { Font, Path, Glyph } from '../src/opentype.js';
+import Layout  from '../src/layout.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/opentypeSpec.js:
```diff
@@ -1,9 +1,11 @@
 import assert from 'assert';
-import { Font, Glyph, load, loadSync, Path } from '../src/opentype';
+import { Font, Path, Glyph, parse, load} from '../src/opentype.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/parse.js:
```diff
@@ -1,6 +1,6 @@
-import { unhex } from './testutil';
-import { Parser } from '../src/parse';
+import { unhex } from './testutil.js';
+import { Parser } from '../src/parse.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/substitution.js:
```diff
@@ -1,6 +1,6 @@
-import { Font, Glyph, Path } from '../src/opentype';
-import Substitution from '../src/substitution';
+import { Font, Glyph, Path } from '../src/opentype.js';
+import Substitution from '../src/substitution.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/table.js:
```diff
@@ -1,6 +1,6 @@
-import { unhexArray } from './testutil';
-import table from '../src/table';
+import { unhexArray } from './testutil.js';
+import table from '../src/table.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/tables/cff.js:
```diff
@@ -1,10 +1,12 @@
+import glyphset from '../../src/glyphset.js';
+import Path from '../../src/path.js';
+import cff from '../../src/tables/cff.js';
+import { parse } from '../../src/opentype.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/tables/colr.js:
```diff
@@ -1,6 +1,6 @@
-import { hex, unhex } from '../testutil';
-import colr from '../../src/tables/colr';
+import { hex, unhex } from '../testutil.js';
+import colr from '../../src/tables/colr.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/tables/cpal.js:
```diff
@@ -1,6 +1,6 @@
-import { hex, unhex } from '../testutil';
-import cpal from '../../src/tables/cpal';
+import { hex, unhex } from '../testutil.js';
+import cpal from '../../src/tables/cpal.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/tables/fvar.js:
```diff
@@ -1,6 +1,6 @@
-import { hex, unhex } from '../testutil';
-import fvar from '../../src/tables/fvar';
+import { hex, unhex } from '../testutil.js';
+import fvar from '../../src/tables/fvar.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/tables/gdef.js:
```diff
@@ -1,6 +1,6 @@
-import { unhex } from '../testutil';
-import gdef from '../../src/tables/gdef';
+import { unhex } from '../testutil.js';
+import gdef from '../../src/tables/gdef.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/tables/gpos.js:
```diff
@@ -1,6 +1,6 @@
-import { unhex } from '../testutil';
-import gpos from '../../src/tables/gpos';
+import { unhex } from '../testutil.js';
+import gpos from '../../src/tables/gpos.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/tables/gsub.js:
```diff
@@ -1,6 +1,6 @@
-import { unhex, unhexArray } from '../testutil';
-import gsub from '../../src/tables/gsub';
+import { unhex, unhexArray } from '../testutil.js';
+import gsub from '../../src/tables/gsub.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/tables/loca.js:
```diff
@@ -1,6 +1,6 @@
-import { unhex } from '../testutil';
-import loca from '../../src/tables/loca';
+import { unhex } from '../testutil.js';
+import loca from '../../src/tables/loca.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/tables/ltag.js:
```diff
@@ -1,6 +1,6 @@
-import { hex, unhex } from '../testutil';
-import ltag from '../../src/tables/ltag';
+import { hex, unhex } from '../testutil.js';
+import ltag from '../../src/tables/ltag.js';
```
Why are we adding js to the end of everything?

yne replied:
to make it work on all platform natively (web browser import don't automatically append .js)

ILOVEPIE commented on test/tables/name.js:
```diff
@@ -1,8 +1,8 @@
+import table from '../../src/table.js';
+import { encode } from '../../src/types.js';
+import _name from '../../src/tables/name.js';
+import { hex, unhex } from '../testutil.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/tokenizer.js:
```diff
@@ -1,6 +1,6 @@
-import Tokenizer, { Token } from '../src/tokenizer';
-import { isWhiteSpace, isArabicChar } from '../src/char';
+import Tokenizer, { Token } from '../src/tokenizer.js';
+import { isWhiteSpace, isArabicChar } from '../src/char.js';
```
Why are we adding js to the end of everything?

ILOVEPIE commented on test/types.js:
```diff
@@ -1,6 +1,6 @@
-import { hex, unhex } from './testutil';
-import { decode, encode, sizeOf } from '../src/types';
+import { hex, unhex } from './testutil.js';
+import { decode, encode, sizeOf } from '../src/types.js';
```
Why are we adding js to the end of everything?

ILOVEPIE reviewed (changes requested):
This should probably be broken up into smaller commits, there's also some changes here that were not properly thought through. Honestly this would have been a lot easier to review if it was two PRs, a build system update and one that adds the github actions.

ILOVEPIE commented on externs/opentype.js:
```diff
@@ -1,651 +0,0 @@
-/**
```
@yne this one's an important one.

yne replied:
I removed it because:
- It is not documented
- It is not used/referenced by any config

I'll add it back but:
- Isn't it duplicate of existing minifier/builder/linter tools ?
- If it's better, why it is not used by default ?
- Why does it duplicate all opentype.js prototypes ? Can't we automatize it ?

ILOVEPIE replied:
@yne btw, that benchmark i think has it running in simple optimizations mode, advanced optimizations mode is even better, but requires JSDoc documentation of all code OR an extern definition file for external libraries that are used (like opentype.js)

yne replied:
Interesting, I would prefer to document the whole source (and enforce it in CI)
so we could avoid repeating public prototypes in /external/opentype.js

... but that's a subject for another PR ;)

yne replied:
I just run `eslint` with `jsdoc` enforcement, and I got 1819 errors ... (missing type, returns, description ...) lot of work to do ...

ILOVEPIE replied:
@yne just a note, closure compiler uses a specific sub-set of jsdoc syntax, so you might want to look at their documentation if you're modifying the /externs/opentype.js file. Closure compiler basically requires externs for all libraries that are loaded separately so it knows what it can and can't optimize. Also, personally i think the best solution would be to annotate all public APIs as such in JSDoc using the  ``@public`` and ``@private`` annotations, and then generate the extern stubs programmatically based on those JSDocs when we tag a release (i.e. create a stub for everything that is marked with ``@public``), but that should probably be a separate PR. I can help with that one as I have some experience with closure compiler.

yne replied:
That would be a very nice improvement.

We could also automatically generate the API documentation and examples of any `@public` methods.

ILOVEPIE commented on package.json:
```diff
@@ -1,8 +1,7 @@
   "description": "OpenType font parser",
-  "version": "2.0.0",
-  "sideEffects": false,
+  "version": "1.3.3",
```
@yne we're working on version 2.0.0 right now, 1.3.3 released a long time ago.

yne replied:
its a mistake from me.
the version shall be left empty/null since it will be set by the github action upon release

yne replied:
I changed to the original name `opentype.js`.
Beware that it will:
- break my repo when I create a tag (since I can't publish `opentype.js` on NPM)
- break this PR if we don't setup an `NPM_TOKEN`

ILOVEPIE replied:
Did we add back in the ``"sideEffects": false,`` entry? It's apparently necessary for tree shaking.

yne replied:
@ILOVEPIE 
it's a webpack-only specific flags (not ESBuild)

yne commented on package.json:
```diff
@@ -1,8 +1,7 @@
 {
-  "name": "opentype.js",
+  "name": "opentype.yne.js",
```
oh good catch !

yne commented on docs/index.html:
```diff
@@ -1,10 +1,9 @@
 <html>
 <head>
-    <title>opentype.js – JavaScript parser/writer for OpenType and TrueType fonts.</title>
+    <title>opentype.js - JavaScript parser/writer for OpenType and TrueType fonts.</title>
```
VSCode editor report it as confusing character.
![image](https://user-images.githubusercontent.com/5113053/220314803-ca018b12-6395-4a93-a863-a9f1120e8fa6.png)

But I'll revert to the original, no problem.

yne commented:
> This should probably be broken up into smaller PRs

Sorry about that, I wanted the jobs to run without vulnerabilities, which lead me to updating our buildtools/testtool, which lead me to fix many lint warning etc...

This is why I carefully split my commit by subjects (action, lint, ...) : to ease the reading (I hope you didn't read this PR as a single page ...).

Also avoid repeating the same comment:
- It's a pain for you, who did a lot of work.
- It's a pain for me, who have to focus on the important one, and close all duplicate after the first explanation.
- It's a pain for others reviewer who need to browse the ~6 subject you raised.

I hope I answered all your question and fix all the changed you asked.
I you spot some remaining mistakes, let me know :+1:

ILOVEPIE commented:
Yeah sorry about the duplicated messages, I was trying an integrated PR reviewing tool in my code editor, I guess it applies the message to everywhere it finds a match if you use it during a code-search.

yne commented:
rebased over https://github.com/opentypejs/opentype.js/pull/566 changes

@ILOVEPIE @Connum do you want/need any other changes ?

ILOVEPIE commented on package.json:
```diff
@@ -1,8 +1,6 @@
   "name": "opentype.js",
   "description": "OpenType font parser",
-  "version": "2.0.0",
-  "sideEffects": false,
```
we need the ``"sideEffects": false,`` entry to allow bundlers to optimize the package.

yne replied:
I thought it was a webpack-specific flag (which is not used anymore)

I've added it back :+1:

Connum replied:
I can't find where the version and side effects properties went?

yne replied:
the version is set by github action upon git tag
See https://github.com/yne/opentype.yne.js/blob/master/.github/workflows/release.yml#L22

the name is also automatically so anybody can fork and rename they fork so they can publish they own fork without conflicting when PR

Connum replied:
But the sideEffects?

yne replied:
it shall be before eslintConfig
it is not ?

Connum replied:
Oh boy... Yes of course it is. I must have missed it in the compact diff view, but I can see it when I open the full package json file. Sorry for the confusion!

ILOVEPIE reviewed (changes requested):
Just need a little clarification on some stuff and one small change to package.json

ILOVEPIE reviewed (approved):
Looks good to me.

Connum commented:
Before someone merges this, give me a second please, I want to add some things!

yne commented:
@Connum no problem, do this merging when you are ready/done :+1: 

@Jolg42 You'll have to setup a Github Action "Repository secrets" named `NPM_TOKEN` to make the npm publish work

Connum reviewed (changes requested):
I didn't have the time for a thorough test, working on too many other things... But I did quickly check this out and try it. Here is something I stumbled upon:

1. I was a bit confused when running `npm run start` and instead of getting the dev server with the index.html, which I frequently use during development
2. So, I added the parameter `--servedir=\"./\"` to the start script, then we can open the pages at http://127.0.0.1:8000/docs/
3. However, we have to change the paths to the js file in the html files as well, `src="dist/opentype.js"` to `src="../dist/opentype.js"` (plus one level down for the files in */docs/examples*). And we also need to add the parameter ` --global-name=opentype` to the script in order to have the opentype global available.
4. Unfortunately, esbuild doesn't support HMR (see evanw/esbuild/issues/796) ☹ But at least we can add the parameters `--watch  --footer:js=\"window.IS_DEV_BUILD=true\"` parameter and add this to the html files:
```js
if(window.IS_DEV_BUILD) {
    new EventSource('/esbuild').addEventListener('change', () => location.reload());
}
```
Now, the page will reload on file changes, which is good enough for me now. So, the final start script should look like this:
```
"start": "esbuild --bundle src/opentype.js --outdir=dist --external:fs --target=es2018 --watch --serve --servedir=\"./\"",
```

If you prefer, I can do those changes and add them to the PR, just let me know.

We must also not forget to set the GitHub pages root to the /docs folder!

But I have to say, I love how blazing fast the build process of esbuild is compared to even vite, which I mainly use for my projects right now. That's really great!

yne commented:
glad you found the solutions :)

> the test command should run the build and dist before running mocha

yes some test rely on /dist/ files
I can't find a way to make an npm script (test) depend on another npm script (dist)

Tips : simply `import` the `./src/opentype.js` work out of the box

for the HMR:
I would prefer doing the reload code directly in the footer:js= instead of imbedding the code in the source behind a flag

I can test your changes tomorrow if you want. or just merge it now and we will see later. as you want

Connum commented:
> I can't find a way to make an npm script (test) depend on another npm script (dist)

Shouldn't `"test": "npm run build && npm run ist && mocha --require reify --recursive",` work?
And in the current setup we have the linting run automatically after testing,  which could otherwise be forgotten if you have to call it manually, so adding `&& npm run lint` could be a good idea as well?

yne commented:
> > I can't find a way to make an npm script (test) depend on another npm script (dist)
> 
> Shouldn't `"test": "npm run build && npm run ist && mocha --require reify --recursive",` work? And in the current setup we have the linting run automatically after testing, which could otherwise be forgotten if you have to call it manually, so adding `&& npm run lint` could be a good idea as well?

Good point, I'll change that :+1:

yne commented:
rebased over https://github.com/opentypejs/opentype.js/pull/575 
@Connum I updated the code according to the discussion
- `npm run test` now also build dev + prod dist files
- `npm run start` now serve the global file + auto-reload
- `npm run lint` now fail upon fixable errors
- README.md was edited to reflect those changes

Tell me if I missed anything

Connum commented on package.json:
```diff
@@ -22,50 +20,64 @@
-    "node": ">= 8.0.0"
+    "build": "npm run b:umd && npm run b:esm",
+    "dist": " npm run d:umd && npm run d:esm",
+    "test": "npm run build && npm run dist && mocha --require reify --recursive",
```
Didn't we want to add lining after the testing?

yne replied:
you want `npm run test` to do the linting ?

Connum replied:
Right now we run linting and coding style at the end of the test script

yne replied:
I would have prefered to keep them separated (CI will check both anyway) but if that's how it was before... I'll add it back

Connum replied:
I'd like running them before I commit or push anything, and knowing myself I'll surely forget the linting :)

yne replied:
I'll leave the `lint` target since it's a burden to re-run all unit test only to get the linting error

Connum replied:
Absolutely! And we don't need to run linting twice during CI, so it just needs to call the test script

yne replied:
Done :heavy_check_mark:

yne replied:
The `version` is set at publication 
see https://github.com/yne/opentype.yne.js/blob/master/.github/workflows/release.yml#L22

yne replied:
the `version` is set by github action upon git tag
See https://github.com/yne/opentype.yne.js/blob/master/.github/workflows/release.yml#L22

the `name` is also automatically so anybody can fork and rename they fork so they can publish they own fork without conflicting when PR

Connum reviewed (approved):
For some reason, the page reload on rebuild does not get triggered...
And I have to double-escape all line breaks in the script parameters to `\\n` in order for the scripts to work.
We can fix that later in order to get this finally over with, but could you please check if the commands still work for you with double-encoded line breaks?

yne commented:
> For some reason, the page reload on rebuild does not get triggered...

Fixed, see https://github.com/opentypejs/opentype.js/pull/578

> And I have to double-escape all line breaks in the script parameters to `\\n` in order for the scripts to work. We can fix that later in order to get this finally over with, but could you please check if the commands still work for you with double-encoded line breaks?

No issue on Linux, but I can remove all lines if it breaks windows build

Connum commented:
I just noticed that the font inspector and glyph inspector don't work, while the index page does: https://opentype.js.org/index.html
I'm quite sure it all worked when I tested it locally while reviewing, but I'm not able to test that right now

yne commented:
The pages are working, but they "demo" font is not loaded (404)
- index load `FiraSansMedium.woff` by default
- while other pages load `Roboto-Black.ttf` by default

I'll unify them all to `Roboto-Black.ttf`

yne commented:
>  we'll have to get rid of the line breaks!

done: 234a535

ILOVEPIE commented:
Sorry been out of the loop a little bit as i've been a bit busy. I'll spend tommorow trying to catch up.

yne commented:
> Great! Can you add it to master?

done
